### PR TITLE
swaync: add module

### PR DESCRIPTION
**Pull request by rhoriguchi**
### Description

Add module `services.swaync`

### Checklist

- [X] Change is backwards compatible.

- [X] Code formatted with `./format`.

- [ ] Code tested through `nix-shell --pure tests -A run.all` or `nix develop --ignore-environment .#all` using Flakes.

- [ ] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [X] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [X] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

### CODE REVIEW

File: modules/lib/maintainers.nix [hunk 1]
```diff
@@ -363,6 +363,12 @@
       fingerprint = "D5D6 FD1F 0D9A 3284 FB9B  C26D 3F98 EC7E C2B8 7ED1";
     }];
   };
+  rhoriguchi = {
+    email = "ryan.horiguchi@gmail.com";
+    github = "rhoriguchi";
+    githubId = 6047658;
+    name = "Ryan Horiguchi";
+  };
```
Comment (rycee) [inline modules/lib/maintainers.nix:371]: No need to add yourself if you are already in Nixpkgs' maintainer list.

File: modules/services/swaync.nix [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    enable = lib.mkEnableOption "Swaync notification daemon";
+
+    package = lib.mkOption {
+      type = lib.types.package;
+      default = pkgs.swaynotificationcenter;
+      defaultText = lib.literalExpression "pkgs.swaynotificationcenter";
+      description = ''
+        Swaync package to use. Set to <code>null</code> to use the default package.
+      '';
+    };
```
Comment (rycee) [inline modules/services/swaync.nix:30]: Can simplify:

```suggestion
    package = lib.mkPackageOption pkgs "swaynotificationcenter" { };
```

File: modules/services/swaync.nix [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        BusName = "org.freedesktop.Notifications";
+        ExecStart = "${cfg.package}/bin/swaync";
+        ExecReload = [
+          "${cfg.package}/bin/swaync-client --reload-config"
+          "${cfg.package}/bin/swaync-client --reload-css"
+        ];
+        Restart = "on-failure";
+      };
+
+      Install.WantedBy = [ cfg.systemd.target ];
```
Comment (rycee) [inline modules/services/swaync.nix:151]: This should not be necessary, the service should get activated automatically by dbus:

```suggestion
```

File: modules/services/swaync.nix [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+      target = lib.mkOption {
+        type = lib.types.str;
+        default = "graphical-session.target";
+        example = "sway-session.target";
+        description = ''
+          Systemd target to bind to.
+        '';
+      };
+    };
+
```
Comment (rycee) [inline modules/services/swaync.nix:44]: I would suggest always creating the systemd service.
```suggestion
```

File: modules/services/swaync.nix [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+      default = null;
+      description = ''
+        CSS style of the bar.
+        </para>
+        <para>
+        See <link xlink:href="https://github.com/ErikReider/SwayNotificationCenter/blob/main/src/style.css"/>
+        for the documentation.
+        </para>
+        <para>
+        If the value is set to a path literal, then the path will be used as the css file.
```
Comment (rycee) [inline modules/services/swaync.nix:56]: Should be Markdown now.

```suggestion
        CSS style of the bar. See
        <https://github.com/ErikReider/SwayNotificationCenter/blob/main/src/style.css>
        for the documentation.

        If the value is set to a path literal, then the path will be used as the CSS file.
```

File: modules/services/swaync.nix [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    };
+
+    settings = lib.mkOption {
+      type = jsonFormat.type;
+      default = { };
+      description = ''
+        Configuration written to <filename>$XDG_CONFIG_HOME/swaync/config.json</filename>.
+        </para>
+        <para>
+        See <link xlink:href="https://github.com/ErikReider/SwayNotificationCenter/blob/main/src/configSchema.json"/>
```
Comment (rycee) [inline modules/services/swaync.nix:85]:
```suggestion
        Configuration written to {file}`$XDG_CONFIG_HOME/swaync/config.json`.

        See <https://github.com/ErikReider/SwayNotificationCenter/blob/main/src/configSchema.json>.
```

File: modules/services/swaync.nix [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+          inherit onChange;
+        };
+      };
+
+    systemd.user.services.swaync = lib.mkIf cfg.systemd.enable {
+      Unit = {
+        Description = "Swaync notification daemon";
+        Documentation = "https://github.com/ErikReider/SwayNotificationCenter";
+        PartOf = [ "graphical-session.target" ];
+        After = [ "graphical-session.target" ];
```
Comment (rycee) [inline modules/services/swaync.nix:136]:
```suggestion
        After = [ "graphical-session-pre.target" ];
```

File: modules/services/swaync.nix [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+          inherit onChange;
+        };
+        "swaync/style.css" = lib.mkIf (cfg.style != null) {
+          source =
+            if builtins.isPath cfg.style || lib.isStorePath cfg.style then
+              cfg.style
+            else
+              pkgs.writeText "swaync/style.css" cfg.style;
+          inherit onChange;
+        };
```
Comment (rycee) [inline modules/services/swaync.nix:128]: I think you can remove the onChange stuff. Just add

```
Unit.X-Restart-Triggers = [
  "${config.xdg.configFile."swaync/config.json".source}"
  "${config.xdg.configFile."swaync/style.css".source}"
]
```

to the service definition.

File: modules/services/swaync.nix (hunk 4 above)
Comment (rhoriguchi) [reply modules/services/swaync.nix:37]: My thought process was, if you only want to use the notification shade but not the notification popups. This way you can configure it without starting it.

File: modules/services/swaync.nix (hunk 3 above)
Comment (rhoriguchi) [reply modules/services/swaync.nix:135]: I need this for the same reason as I need it here. https://github.com/nix-community/home-manager/pull/4224#issuecomment-1631526382

File: modules/services/swaync.nix (hunk 5 above)
Comment (rhoriguchi) [reply modules/services/swaync.nix]: I had to remove the `<` and `>` from the url to get it to build.

File: modules/services/swaync.nix (hunk 8 above)
Comment (rhoriguchi) [reply modules/services/swaync.nix]: There's a couple of other places where this improvement could be made.

**Comment by rhoriguchi** [comment]
Thx for the inputs @rycee

**Comment by teto** [comment]
I am interested in this as a swaync user but this breaks my nixos-unstable build: 
```
    > /nix/store/lsg1z6y6k5jp5wz2w0yqf8zs92n62j1y-manual-combined/man
ual-combined.xml:51457: validity error : xml:id : attribute value opt-services.swaync.settings._\$schema_ is not an NCName
```
also could you rebase ? other than that looks ok

File: modules/services/swaync.nix (hunk 4 above)
Comment (teto) [reply modules/services/swaync.nix:33]: One could still disable the systemd in its own config via systemd.user.services.swaync.enable = false or something similar (not sure how to disable the service) , i.e., it makes the module simpler.

**Comment by teto** [comment]
I've been using this for a while and only now realized it was not merged yet. Could you rebase and address comments please ? it's a pretty cool software/module

**Comment by rhoriguchi** [comment]
I'm currently not running hyperland, so I can't really test. Thank you @teto for rebasing.

**Comment by hmajid2301** [comment]
hi guys any updates on this ?

**Comment by teto** [comment]
@hmajid2301 feel free to open a new PR with this as a basis as long as you keep the commit authors intact

**Comment by mobsenpai** [comment]
Looks like only formatting tests are failing, isn't it?

**Comment by abayomi185** [comment]
I'd like to look into this. Will create a PR referencing this

**Comment by rycee** [comment]
Merged #5345 so I'll close this. Thanks everybody for your input!

**Pull request closed without merging**